### ISSUE TITLE
`ed*` scripts fail with rootless Docker due to `docker0` interface not existing

### ISSUE THREAD

**Issue by sergiofenoll** [madnificent/docker-ember#36]
If you run the provided scripts on a system using rootless Docker they will complain about the `docker0` interface not existing and then fail.

```
Device "docker0" does not exist.
invalid argument "host:" for "--add-host" flag: invalid IP address in add-host: ""
See 'docker run --help'.
```
This occurs when calculating `DOCKERHOST` in `edi_docker_host_option`:
```
if [[ "$OSTYPE" == "linux-gnu" || "$OSTYPE" == "linux" ]]
then
    # Linux
    DOCKERHOST=`ip addr show docker0 | grep inet | head -n 1 | awk '{ print $2 }' | grep -oP "^[^/]+"`
# [ ... ]
fi
echo "--add-host host:$DOCKERHOST "
```
[This issue](https://github.com/moby/moby/issues/43116) seems to point to it being intentional.

~~The rest of the function links to a SO question, the accepted answers suggests using `--add-host host.docker.internal:host-gateway` for more modern Docker installs (> v20.10.0). If I change the scripts to use `host.docker.internal:host-gateway` they work, but I'm not sure if it's compatible with root Docker installs and/or other OSes. So I don't know if we want to change this for across the board. But it may be useful for anyone else using rootless Docker.~~
EDIT: actually, using the MacOS command worked. Apparently rootless Docker does not let containers communicate with the host via the docker0 interface. So inside the ed containers `host` needs to point to my computer's network interface's IP.
EDIT 2: When offline, this won't work, because it's trying to get the IP address of your network interface, which AFAIK isn't set when not connected to a network. I assume it's also the case for MacOS users, but I'm not sure how we could work around this.

**Comment by piemonkey** [comment]
I just put in a PR for this which uses `nsenter` to find the IP: #43. Are you able to verify if this works for you? I've tested it, but only under Arch and using the `docker-rootless-extras` package, so maybe it's different on other distros.

**Comment by sergiofenoll** [comment]
> I just put in a PR for this which uses `nsenter` to find the IP: #43. Are you able to verify if this works for you? I've tested it, but only under Arch and using the `docker-rootless-extras` package, so maybe it's different on other distros.

Thanks for the PR @piemonkey! But I can't get it to work on my machine. Though using `nsenter` an IP for the `docker0` interface is found (`172.17.0.1` in my case) the `eds` container can't actually use it to communicate with the other containers. All requests result in the following error:

![image](https://github.com/madnificent/docker-ember/assets/22411874/578de49a-ad2b-49ee-8446-5970fb79383e)

It doesn't seem related to exposing privileged ports (that should already be set up as I can navigate to `http://localhost` from my browser and work with the running identifier that's bound to said port.

BTW, what I ended up doing to make `eds` work locally is invoking it like this:

```
eds --proxy=http://(hostname)  # (hostname) is fish syntax, on bash it should be $(hostname) or `hostname`
```

EDIT: all this is using Fedora Silverblue

**Comment by piemonkey** [comment]
Hmm, actually after restarting my machine I get this too. I guess that's what I get for playing around with rootless and with-root docker at the same time.

I don't get how `$(hostname)` works for you as how does the container know what that name refers to. Unless it's something about how your local network is set up, so that it maps hostnames to hosts within the network. For me I just get `getaddrinfo ENOTFOUND` for the string version of my hostname.

I've realised a better way to do this with docker networking but it'll take some tweaking to the API of `eds` to use it, so I'll have to get to that later.

**Comment by sergiofenoll** [comment]
I can't say for sure why (on my machine, not on my network) my machine's hostname resolves to my own machine. As far as I remember I didn't have to do anything special to set it up, it just worked like this out of the box. Perhaps it's related to some configuration in `systemd-resolved`?

Looking forward to seeing the better way of networking you've found!

**Comment by piemonkey** [comment]
I've finally got round to committing it and I've updated my PR #43 

Now you can use e.g. `eds --network=mu-project_default --proxy http://identifier` to proxy correctly.

**Comment by erikap** [comment]
Fixed by #43
